### PR TITLE
Restrict the allowed types of path parameters #9

### PR DESCRIPTION
**Pull request by harshilparmar**
## Summary

I have made pr for #9 .Which will restrict user to pass not allowed value to package function.

## Details

- [x] Made certain change to restrict types
- [x]  Test cases

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -37,13 +37,18 @@ export function subst(template: string, params: ParamMap) {
...
 
   const renderedPath = template.replace(/:\w+/g, p => {
     const key = p.slice(1);
     if (!params.hasOwnProperty(key)) {
       return p;
     }
     delete remainingParams[key];
+
+    if (!allowedTypes.includes(typeof params[key])) {
+      throw new TypeError(`Path parameter ${key} cannot be of type ${typeof params[key]}.Allowed types are: ${allowedTypes.join(', ')}.`);
```
Comment (balazsbotond) [inline src/index.ts]: This is just cosmetics, but a space is missing before the word "Allowed".

**Comment by balazsbotond** [comment]
The code looks good to me! Looking forward to the test cases.

**Comment by harshilparmar** [comment]
> 
> 
> The code looks good to me! Looking forward to the test cases.

Bro,For which types I have to write test cases? only for allowedTypes?

**Comment by balazsbotond** [comment]
For allowed types -> the path should be rendered
For symbol, array and object -> a TypeError should be thrown

Null and undefined are removed before the path call so no tests are needed for those.

BigInt isn't available in most environments so I'd skip testing it.

**Comment by harshilparmar** [comment]
@balazsbotond Done 🤗  ,Can you please check !!

File: test/urlcat.ts [hunk 2]
```diff
@@ -3,79 +3,79 @@ import { expect } from 'chai';
...
     const expected = 'http://example.com/path/1?q=2';
     const actual = urlcat('http://example.com/', '/path/:p', { p: 1, q: 2 });
     expect(actual).to.equal(expected);
   });
 
-  it ('Uses exactly one ? to join query parameters even if the path has a trailing question mark', () => {
-    const expected = 'http://example.com/path?q=2';
-    const actual = urlcat('http://example.com/', '/path?', { q: 2 });
+  it('Uses exactly one ? to join query parameters even if the path has a trailing question mark', () => {
+    const expected = 'http://example.com/path?q=true';
```
Comment (balazsbotond) [inline test/urlcat.ts]: Why did you change `q` to `true` here?

File: test/urlcat.ts [hunk 3]
```diff
@@ -3,79 +3,79 @@ import { expect } from 'chai';
 
 describe('urlcat', () => {
 
-  it ('Concatenates the base URL and the path if no params are passed', () => {
+  it('Concatenates the base URL and the path if no params are passed', () => {
```
Comment (balazsbotond) [inline test/urlcat.ts:6]: Good idea to remove the extra space here - I haven't noticed it. Thanks!

File: test/urlcat.ts [hunk 4]
```diff
@@ -85,22 +85,71 @@ describe('urlcat', () => {
...
+  });
+
+  it('Can handle allowed Type String as param', () => {
+    const expected = 'http://example.com/path/test';
+    const actual = urlcat('http://example.com/path/:p', { p: 'test' });
+    expect(actual).to.equal(expected);
+  });
+
+  it('Can\'t handle object as param', () => {
+    expect(() => urlcat('http://example.com/path/:p', { p: {} })).to.throw(TypeError, "Path parameter p cannot be of type object.Allowed types are: boolean, string, number.");
```
Comment (balazsbotond) [inline test/urlcat.ts]: Lines like this are a bit too long to be readable. I would suggest inserting a line break before `.to.throw(...`.

File: test/urlcat.ts [hunk 5]
```diff
@@ -85,22 +85,71 @@ describe('urlcat', () => {
...
+
+  it('Can\'t handle array as param', () => {
+    expect(() => urlcat('http://example.com/path/:p/:q', { p: [] })).to.throw(TypeError, "Path parameter p cannot be of type object.Allowed types are: boolean, string, number.");
+  });
+
+  it('Can\'t handle symbol as param', () => {
+    expect(() => urlcat('http://example.com/path/:p', { p: Symbol() })).to.throw(TypeError, "Path parameter p cannot be of type symbol.Allowed types are: boolean, string, number.");
+  });
+
+  it('Can\'t handle null as param', () => {
```
Comment (balazsbotond) [inline test/urlcat.ts]: Could you please create a separate PR for #12? I have some comments related to those changes; this way I could merge this PR more quickly and your work would show up as two separate contributions :)

File: src/index.ts [hunk 6]
```diff
@@ -68,5 +73,7 @@ function removeNullOrUndef(params: ParamMap) {
 }
 
 function notNullOrUndefined(v: any) {
+  if (v === null || v === undefined) throw new TypeError(`Path parameter ${v} is not Valid`);
+
```
Comment (balazsbotond) [inline src/index.ts]: Could you please create a separate PR for #12? I have some comments related to those changes; this way I could merge this PR more quickly and your work would show up as two separate contributions :)

**Review by balazsbotond — CHANGES_REQUESTED** [review]
Thank you! The changes related to #9 are correct. I requested few little things here and there, but after you fix those, I will merge the PR.

Could you please create a separate PR for #12? I have some comments related to those changes; this way I could merge this PR more quickly and your work would show up as two separate contributions :)

File: test/urlcat.ts (hunk 2 above)
Comment (harshilparmar) [reply test/urlcat.ts]: ohh, sorry for that,I was just trying but forgot to delete.

**Comment by harshilparmar** [comment]
@balazsbotond I have made this pr for issue #9,and I will make another pr for #13.

**Comment by balazsbotond** [comment]
Yes, you're right, this is #12

**Review by balazsbotond — APPROVED** [review]
Thank you, and congrats on creating the first code PR on this project!

**Comment by harshilparmar** [comment]
@balazsbotond Thanks!! Can I give a suggestion ? How about making GitHub organization for this project and future ones.
